L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+#include "L1Trigger/L1THGCal/interface/backend/HGCalShowerShape.h"
+#include "L1Trigger/L1THGCal/interface/backend/HGCalClusteringDummyImpl.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+// #include "Geometry/HcalCommonData/interface/HcalDDDRecConstants.h"
+// #include "Geometry/HcalCommonData/interface/HcalHitRelabeller.h"
+// #include "SimDataFormats/CaloTest/interface/HGCalTestNumbering.h"
+
+class CaloTruthCellsProducer : public edm::stream::EDProducer<> {
+public:
+  explicit CaloTruthCellsProducer(edm::ParameterSet const&);
+  ~CaloTruthCellsProducer();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  virtual void beginStream(edm::StreamID) override;
+  virtual void produce(edm::Event&, edm::EventSetup const&) override;
+  virtual void endStream() override;
+
+  virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  virtual void endRun(edm::Run const&, edm::EventSetup const&) override;
+  virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+  virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+
+  // std::map<uint32_t, double> makeHitMap(edm::Event const&, edm::EventSetup const&) const;
+
+  typedef edm::AssociationMap<edm::OneToMany<CaloParticleCollection, l1t::HGCalTriggerCellBxCollection>> CaloToCellsMap;
+
+  bool makeCellsCollection_;
+  edm::EDGetTokenT<CaloParticleCollection> caloParticlesToken_;
+  edm::EDGetTokenT<l1t::HGCalTriggerCellBxCollection> triggerCellsToken_;
+  // // three simhits tokens, depedent on old or new geometry
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenEE_;
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEfront_;
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEback_;
+
+  // // for input with old (<= V8) geometry, sim and reco hits use different ID schemes
+  // bool needSimToRecoMapping_{false};
+
+  HGCalClusteringDummyImpl dummyClustering_;
+  HGCalShowerShape showerShape_;
+};
+
+CaloTruthCellsProducer::CaloTruthCellsProducer(edm::ParameterSet const& _config) :
+  makeCellsCollection_(_config.getParameter<bool>("makeCellsCollection")),
+  caloParticlesToken_(consumes<CaloParticleCollection>(_config.getParameter<edm::InputTag>("caloParticles"))),
+  triggerCellsToken_(consumes<l1t::HGCalTriggerCellBxCollection>(_config.getParameter<edm::InputTag>("triggerCells"))),
+  // simHitsTokenEE_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsEE"))),
+  // simHitsTokenHEfront_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEfront"))),
+  // simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEback")))
+  dummyClustering_(_config.getParameterSet("dummyClustering"))
+{
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>("HGCalBackendLayer1Processor2DClustering");
```

Comment:
I thought using the standard collection names would make integration into hgCalTriggerChains easier. @jbsauvan how do you foresee the chain integration?

Reply:
There are two ways the truth matching can be used:
- Running sim clustering on trigger cells or SuperTC to evaluate the frontend algorithms (by the way it currently doesn't work on STC; for that it would require some changes in the way STC are built)
- Running standard clustering on sim-matched trigger cells to evaluate clustering algorithms

So I would see 2 additional methods in the `HGCalTriggerChains` that would create such special chain.

Reply:
OK sounds good!